### DISCUSSION TITLE
`sort` + `uniq` is super slow?

### DISCUSSION THREAD

**Discussion by davidatsurge** [nushell/nushell#7691] [Q&A] [answered]
With [this CSV file](https://raw.githubusercontent.com/nytimes/ingredient-phrase-tagger/master/nyt-ingredients-snapshot-2015.csv) downloaded, the following nushell command takes forever (I haven't had the patience to time it, but at least on the order of minutes):

```
open nyt-ingredients-snapshot-2015.csv | select name | sort | uniq 
```

While doing the following in bash takes ~1 sec.

```bash
xsv select name nyt-ingredients-snapshot-2015.csv  | sort | uniq 
```

`xsv` above is [this tool](https://github.com/BurntSushi/xsv/).

Is this perf disparity to be expected, or am I doing something wrong?

**Comment by hustcer** [comment]
For large dataset It's recommend to use dataframe commands (Use `help commands | where category == dataframe or category == lazyframe` to filter them) such as:
`open nyt-ingredients-snapshot-2015.csv | into df | select name | unique | n-unique`
`open nyt-ingredients-snapshot-2015.csv | into df | select name | unique | sort-by name`

```
@hustcer ➜ tmp git(develop) $> benchmark { open nyt-ingredients-snapshot-2015.csv | into df|select name|unique|n-unique }
1sec 166ms 524µs 671ns
@hustcer ➜ tmp git(develop) $> benchmark { open nyt-ingredients-snapshot-2015.csv | into df|select name|unique|sort-by name }
1sec 159ms 634µs 820ns
```
For more detail: https://www.nushell.sh/book/dataframes.html

**Reply by davidatsurge** [reply]
Thank you very much!

**Comment by hustcer** [comment]
BTW, use `open-df` will be faster:
```
@hustcer ➜ tmp git(develop) $> benchmark { open-df nyt-ingredients-snapshot-2015.csv | select name | unique | n-unique }
54ms 962µs 662ns
@hustcer ➜ tmp git(develop) $> benchmark { open-df nyt-ingredients-snapshot-2015.csv | select name | unique | sort-by name }
48ms 567µs 578ns
```

**Comment by fdncred** [comment, marked as answer]
We've recently made improvements to uniq that are on the latest main branch. This is why I get without dataframes.
```
benchmark { open nyt.csv | select name | sort | uniq }
2sec 49ms 61µs 100ns
```
and with dataframes (although I had to infer the schema)
```
benchmark { open-df nyt.csv --infer-schema 1000 | select name | sort-by name | unique }
40ms 582µs 900ns
```

**Reply by davidatsurge** [reply]
Thank you!
